src/uploaders/BrowserUploader.ts

```diff
@@ -0,0 +1,141 @@
+import { promises as fs } from 'fs'
+import path from 'path'
+import http from 'http'
+import readPkgUp from 'read-pkg-up'
+
+import { Logger, noopLogger } from '../Logger'
+import request from '../Request'
+import { UploadErrorCode, UploadError } from '../UploadError'
+import AddSources from '../transformers/AddSources'
+import StripProjectRoot from '../transformers/StripProjectRoot'
+
+interface UploadOpts {
+  apiKey: string,
+  sourceMap: string,
+  bundleUrl: string,
+  bundle?: string,
+  appVersion?: string,
+  overwrite?: boolean
+  projectRoot?: string
+  endpoint?: string
+  requestOpts?: http.RequestOptions
+  logger?: Logger
+}
+
+export async function uploadOne ({
+  apiKey,
+  bundleUrl,
+  bundle,
+  sourceMap,
+  appVersion,
+  overwrite = false,
+  projectRoot = process.cwd(),
+  endpoint = 'https://upload.bugsnag.com/',
+  requestOpts = {},
+  logger = noopLogger
+}: UploadOpts): Promise<void> {
+  logger.info(`Uploading browser source map for "${bundleUrl}"`)
+
+  logger.debug(`Reading source map "${sourceMap}"`)
+  const fullSourceMapPath = path.resolve(projectRoot, sourceMap)
+  let sourceMapContent
+  try {
+    sourceMapContent = await fs.readFile(fullSourceMapPath, 'utf-8')
+  } catch (e) {
+    logger.error(`There was an error attempting to find a source map at the following location. Is the path correct?\n\n  "${fullSourceMapPath}"`)
+    throw e
+  }
+
+  let bundleContent
+  let fullBundlePath
+  if (bundle) {
+    fullBundlePath = path.resolve(projectRoot, bundle)
+    logger.debug(`Reading bundle file "${bundle}"`)
+    try {
+      bundleContent = await fs.readFile(fullBundlePath, 'utf-8')
+    } catch (e) {
+      logger.error(`There was an error attempting to find a bundle file at the following location. Is the path correct?\n\n  "${fullBundlePath}"`)
+      throw e
+    }
+  }
+
+  let sourceMapJson
+  try {
+    sourceMapJson = JSON.parse(sourceMapContent)
+  } catch (e) {
+    logger.error(`The provided source map was not valid JSON`)
+    throw e
+  }
+
+  logger.info('Applying transformations to source map')
+  let transformedSourceMap
+  try {
+    transformedSourceMap = await Promise.resolve(sourceMapJson)
+      .then(json => AddSources(fullSourceMapPath, json, projectRoot, logger))
+      .then(json => StripProjectRoot(fullSourceMapPath, json, projectRoot, logger))
+  } catch (e) {
+    logger.error('Error applying transforms to source map')
+  }
+
+  if (!appVersion) {
+    logger.debug('Detecting appVersion')
+    appVersion = await detectAppVersion(projectRoot)
```

Comment:
Is it more useful if this also logged the version it detected? It doesn't look like this is displayed elsewhere

```suggestion
    appVersion = await detectAppVersion(projectRoot)
    logger.debug(`Detected appVersion '${appVersion}'`)
```